VHD restructuring

nraynaud opened the pull request:
linked to  #2792

nraynaud committed:
vhd: move remotes code to a separate module

nraynaud committed:
remove streamToBuffer, and cleanup @xen-orchestra/fs utils.js

nraynaud committed:
try to get travis to build the new package

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vhd: mode vhd-merge to new module.

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vhd: re-wire lib-vhd module to dependants

nraynaud committed:
vhd: cleanup, put back flow annotations

nraynaud committed:
vhd: cleanup in remotes

nraynaud committed:
vhd: split a vhd.js file

nraynaud committed:
vhd: wire vhd-cli with vhd-lib

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vhd: convert vmdk-to-vhd tests to jest.

nraynaud committed:
vhd: convert vmdk-to-vhd tests to jest.

nraynaud committed:
vhd: convert vmdk-to-vhd tests to jest.

nraynaud committed:
vhd: run vmdk-to-vhd tests in a tmp directory

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vhd: add some vhd-util checks to vmdk-to-vhd tests

nraynaud committed:
vmdk-to-vhd : use vhd-lib structs for header and footer

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

# Conflicts:
#	packages/xo-server/package.json

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vmdk-to-vhd : use vhd-lib for geometry and a few constants

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

# Conflicts:
#	packages/xo-server/src/api/xo.js
#	yarn.lock

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vmdk-to-vhd : remove unused code

nraynaud committed:
vmdk-to-vhd : use async/await

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vmdk-to-vhd : move all VHD code to vhd-lib

nraynaud committed:
vhd-lib: stop exposing computeGeometryForSize

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vhd-lib: add a fromEvent in tests.

nraynaud commented:
I am ready for review and merging.
I didn't do the vhd-cli changes and the compression, I'd like to chunk the thing.

julien-f commented on .idea/watcherTasks.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectTasksOptions" suppressed-tasks="SCSS" />
+</project>
```
What are these files?

nraynaud replied:
mistake, ide files

julien-f commented on .travis.yml:
```diff
@@ -21,5 +22,6 @@ cache:
   yarn: true
 
 script:
+  - yarn run build
```
Avoid building everything, simply change the script `prepublishOnly` to `prepare` for packages needed for the tests.

julien-f commented on .yarnclean:
```diff
@@ -0,0 +1,42 @@
+
+# misc
+*.gz
+*.md
```
Why this?

nraynaud replied:
no idea, should I remove the entire file?

julien-f replied:
Yes please, avoid adding unnecessary files.

julien-f commented on @xen-orchestra/fs/src/abstract.js:
```diff
@@ -1,11 +1,11 @@
 
+import { randomBytes } from 'crypto'
+import getStream from 'get-stream'
+import { fromCallback, fromEvent, ignoreErrors } from 'promise-toolbox'
```
Sort imports

julien-f commented on @xen-orchestra/fs/src/abstract.js:
```diff
@@ -54,7 +54,7 @@ export default class RemoteHandlerAbstract {
   async test (): Promise<Object> {
     const testFileName = `${Date.now()}.test`
-    const data = getPseudoRandomBytes(1024 * 1024)
+    const data = await fromCallback(cb => randomBytes(1024 * 1024, cb))
```
`randomBytes` is much slower than `getPseudoRandomBytes` (5-10× slower in latest tests), but it probably does not matter for so little data.

nraynaud replied:
getPseudoRandomBytes() is in xo-server, I replaced it with a standard library function instead of copy-pasting it.

julien-f commented on @xen-orchestra/vhd-lib/src/vhd.js:
```diff
@@ -45,28 +53,37 @@ export const PLATFORM_W2KU = 0x57326b75
   fu.uint32('features'), // 8
   fu.uint32('fileFormatVersion'), // 12
-  uint64('dataOffset'), // offset of the header, should always be 512
+  uint64Undefinable('dataOffset'), // offset of the header
```
Can you explain this change?

nraynaud replied:
it's a special 64 integer field that can have an "undefined" value where all its bits are 1. In this instance it's used for HARD_DISK_TYPE_FIXED disk types, who don't have a header.

julien-f replied:
Should we really take care of it?

I think we should explicitly only support dynamic and differencing disks.

nraynaud replied:
VMDK to VHD relies on fixed VHDs.

nraynaud committed:
remove IDE files

nraynaud committed:
avoid global build in travis

nraynaud committed:
import order

nraynaud committed:
vhd-lib: simplify fixed VHD stream generator

nraynaud committed:
vhd-lib: fix padding in VHD stream generator

nraynaud committed:
remove .yarnclean files

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud commented:
STOP THE PRESS !
I made a mistake around vmdk-to-vhd, it doesn't work with xo-web (I had forgotten this is client-side code), and I have an issue with the connection with my VM at the mothership, so I'll fix it tomorrow.

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

julien-f committed:
chore: update yarn.lock

julien-f committed:
chore: upgrade Babel deps

julien-f committed:
chore: remove unused Babel plugins

julien-f committed:
fix: add Babel runtime to new packages

nraynaud committed:
Merge remote-tracking branch 'origin/nr-vhd-restructuring' into nr-vhd-restructuring

nraynaud committed:
vhd-lib: fix vmdk-to-vhd result file

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
vhd-lib: add assert to vhd test

nraynaud committed:
vhd-lib: bump async-iterator-to-stream version

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

julien-f committed:
@xen-orchestra/vhd-lib → vhd-lib

julien-f committed:
split xo-lib in multiple modules

nraynaud committed:
vhd-lib: fix tests

nraynaud committed:
update syntheticstream todo

nraynaud committed:
improve syntheticstream only test

julien-f committed:
Merge branch 'master' into nr-vhd-restructuring

julien-f committed:
remove Vhd#hasBlockAllocationTableMap

julien-f committed:
Vhd: readBlockTable → readBlockAllocationTable

julien-f committed:
assert parent disk type

julien-f committed:
add flow type for promise-toolbox/fromCallback

julien-f committed:
header version

julien-f committed:
mergeVhd: remove irrelevant TODO

julien-f committed:
add minor TODO

julien-f committed:
add another TODO

julien-f committed:
add vhd-cli synthetize

julien-f committed:
add more asserts

julien-f committed:
allow size to be more than a round number of blocks

julien-f committed:
use inequality for maxTableENtries

nraynaud committed:
fix inequality for maxTableEntries

nraynaud committed:
change VMDK import system
 - now extracts the block allocation table client side and send it to
 the server
 - the generated VHD is now sparse

nraynaud committed:
Merge remote-tracking branch 'origin/master' into nr-vhd-restructuring

nraynaud committed:
add test to VMDK import system

nraynaud committed:
VMDK system: fix import issue

julien-f committed:
check second footer

julien-f committed:
assert enough bytes read in Vhd#_read

julien-f committed:
add vhd-cli check

julien-f committed:
remove useless shebang

julien-f committed:
more constants

nraynaud committed:
VMDK system: remove unused feature from virtual-buffer.

nraynaud committed:
VMDK system: remove unused feature from virtual-buffer.

nraynaud committed:
VMDK system: remove unused feature from virtual-buffer.
